### PR TITLE
Add CropYear.dataSourceType and Project.supplierId to v4 spec

### PR DESCRIPTION
**Pull request by ncanning**
NO-2287

Added a new field for Bayer to distinguish between data reported by growers, and projected future data.

### CODE REVIEW

File: packages/project/src/v4-specification.ts [hunk 1]
```diff
@@ -1200,6 +1209,18 @@ export interface CropYear {
...
+   * or a projection of future data.
+   *
+   * @example <caption>When data is from a projection of future data:</caption>
+   *
+   * ```js
+   * "dataTimeFrame": "PROJECTED"
+   * ```
+   *
+   */
+  dataTimeFrame: DataTimeFrame;
```
Comment (ncanning) [inline packages/project/src/v4-specification.ts:1223]: This is the actual change I made

File: packages/project/src/v4-specification.ts [hunk 2]
```diff
@@ -1200,6 +1209,18 @@ export interface CropYear {
     (AnnualCrop | CoverCrop | OrchardOrVineyardCrop | PerennialCrop)?,
     (AnnualCrop | CoverCrop | OrchardOrVineyardCrop | PerennialCrop)?
   ];
+  /**
+   * Enum indicating whether data is historical data reported by the grower,
+   * or a projection of future data.
+   *
+   * @example <caption>When data is from a projection of future data:</caption>
+   *
```
Comment (jaycenhorton) [inline packages/project/src/v4-specification.ts:1217]:
```suggestion
   * @nullable
```

File: packages/project/src/v4-specification.ts (hunk 1 above)
Comment (jaycenhorton) [inline packages/project/src/v4-specification.ts:1223]:
```suggestion
  dataTimeFrame?: DataTimeFrame;
```

File: packages/project/src/utils/convertFromV3ToV4.ts [hunk 3]
```diff
@@ -211,6 +211,7 @@ export const convertV3FieldToV4Field = (v3Field: FieldV3): FieldV4 => {
       return {
         plantingYear: v3CropYear.plantingYear,
         crops: [crops[0], crops[1], crops[2]],
+        dataTimeFrame: undefined,
```
Comment (jaycenhorton) [inline packages/project/src/utils/convertFromV3ToV4.ts:214]:
```suggestion
```

**Review by jaycenhorton — CHANGES_REQUESTED** [review]
see comments

**Review by mrdave-dev — CHANGES_REQUESTED** [review]
Yeah, that's weird we've lost all the documentation in the `.md` files. Maybe this is typically invoked with some flag to output the docs? We should not be losing these docs

**Comment by ncanning** [comment]
Yeah, unclear. I ran `yarn workspace @nori-dot-com/project make:docs`

**Comment by ncanning** [comment]
Okay, think this should be fixed now. @mrdave-dev updated some dependencies and we had to change the format of error messages because they were getting deleted, but everything looks right to me now

File: package.json [hunk 4]
```diff
@@ -77,5 +77,10 @@
         "preset": "ts-jest"
       }
     ]
+  },
+  "dependencies": {
+    "project": "^0.1.6",
+    "typedoc": "0.23.17",
+    "typedoc-plugin-markdown": "3.13.6"
   }
```
Comment (mrdave-dev) [inline package.json:85]: Oops, did I accidentally add this to wrong package.json?

File: packages/project/src/json/v4-specification.json [hunk 5]
```diff
@@ -379,10 +379,7 @@
             "properties": {
                 "date": {
                     "description": "The date the crop event happened (formatted as ISO8061 date: YYYY-MM-DD and YYYY > 2000 and YYYY < 2100).\n\nDates for liming and burning can be approximate or the first day of the crop year.",
-                    "errorMessage": {
-                        "type": "projectDataError:cropEventDateTypeError",
-                        "validationRules": "projectDataError:cropEventDateValidationRuleViolation"
-                    },
+                    "errorMessage": "```js\n{\n\"type\": \"projectDataError:cropEventDateTypeError\",\n\"validationRules\": \"projectDataError:cropEventDateValidationRuleViolation\"\n}\n```",
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:382]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 6]
```diff
@@ -427,10 +424,7 @@
...
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:427]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 7]
```diff
@@ -831,10 +841,7 @@
...
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:844]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 8]
```diff
@@ -921,11 +928,7 @@
         "Field": {
             "additionalProperties": false,
             "description": "A field defining annual crop management practices. Fields are defined by geographic boundaries that contain crop management practices that are identical across the whole of that boundary.",
-            "errorMessage": {
-                "_": "projectDataError:fieldUnknownError",
-                "additionalProperties": "projectDataError:fieldUnknownAdditionalProperty",
-                "required": "projectDataError:fieldRequiredPropertyMissing"
-            },
+            "errorMessage": "```js\n{\n\"required\": \"projectDataError:fieldRequiredPropertyMissing\",\n\"additionalProperties\": \"projectDataError:fieldUnknownAdditionalProperty\",\n\"_\": \"projectDataError:fieldUnknownError\"\n}\n```",
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:931]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 9]
```diff
@@ -967,10 +970,7 @@
                 },
                 "fieldName": {
                     "description": "The name of the field.",
-                    "errorMessage": {
-                        "_": "projectDataError:fieldNameUnknownError",
-                        "type": "projectDataError:fieldNameTypeError"
-                    },
+                    "errorMessage": "```js\n{\n\"type\": \"projectDataError:fieldNameTypeError\",\n\"_\": \"projectDataError:fieldNameUnknownError\"\n}\n```",
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:973]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 10]
```diff
@@ -2025,10 +2025,7 @@
             "properties": {
                 "date": {
                     "description": "The date the crop event happened (formatted as ISO8061 date: YYYY-MM-DD and YYYY > 2000 and YYYY < 2100).\n\nDates for liming and burning can be approximate or the first day of the crop year.",
-                    "errorMessage": {
-                        "type": "projectDataError:cropEventDateTypeError",
-                        "validationRules": "projectDataError:cropEventDateValidationRuleViolation"
-                    },
+                    "errorMessage": "```js\n{\n\"type\": \"projectDataError:cropEventDateTypeError\",\n\"validationRules\": \"projectDataError:cropEventDateValidationRuleViolation\"\n}\n```",
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:2028]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 11]
```diff
@@ -2089,10 +2086,7 @@
...
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:2089]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 12]
```diff
@@ -2351,10 +2345,7 @@
...
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:2348]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 13]
```diff
@@ -2406,10 +2397,7 @@
...
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:2400]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 14]
```diff
@@ -2981,10 +2969,7 @@
...
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:2972]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 15]
```diff
@@ -3079,10 +3064,7 @@
...
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:3067]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 16]
```diff
@@ -3140,10 +3122,7 @@
                 },
                 "date": {
                     "description": "The date the crop event happened (formatted as ISO8061 date: YYYY-MM-DD and YYYY > 2000 and YYYY < 2100).\n\nDates for liming and burning can be approximate or the first day of the crop year.",
-                    "errorMessage": {
-                        "type": "projectDataError:cropEventDateTypeError",
-                        "validationRules": "projectDataError:cropEventDateValidationRuleViolation"
-                    },
+                    "errorMessage": "```js\n{\n\"type\": \"projectDataError:cropEventDateTypeError\",\n\"validationRules\": \"projectDataError:cropEventDateValidationRuleViolation\"\n}\n```",
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:3125]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 17]
```diff
@@ -3242,10 +3221,7 @@
             "properties": {
                 "date": {
                     "description": "The date the crop event happened (formatted as ISO8061 date: YYYY-MM-DD and YYYY > 2000 and YYYY < 2100).\n\nDates for liming and burning can be approximate or the first day of the crop year.",
-                    "errorMessage": {
-                        "type": "projectDataError:cropEventDateTypeError",
-                        "validationRules": "projectDataError:cropEventDateValidationRuleViolation"
-                    },
+                    "errorMessage": "```js\n{\n\"type\": \"projectDataError:cropEventDateTypeError\",\n\"validationRules\": \"projectDataError:cropEventDateValidationRuleViolation\"\n}\n```",
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:3224]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 18]
```diff
@@ -3350,10 +3326,7 @@
                 },
                 "date": {
                     "description": "The date the crop event happened (formatted as ISO8061 date: YYYY-MM-DD and YYYY > 2000 and YYYY < 2100).\n\nDates for liming and burning can be approximate or the first day of the crop year.",
-                    "errorMessage": {
-                        "type": "projectDataError:cropEventDateTypeError",
-                        "validationRules": "projectDataError:cropEventDateValidationRuleViolation"
-                    },
+                    "errorMessage": "```js\n{\n\"type\": \"projectDataError:cropEventDateTypeError\",\n\"validationRules\": \"projectDataError:cropEventDateValidationRuleViolation\"\n}\n```",
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:3329]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 19]
```diff
@@ -3461,9 +3434,7 @@
         }
     },
     "description": "A project encapsulates a set of fields. This is the top-level interface of Nori's Croplands Data Import format.\n\nA project maye represent either a complete farming operation for a single operator or a batch\nof fields from a data aggregator.",
-    "errorMessage": {
-        "_": "projectDataError:projectUnknownError"
-    },
+    "errorMessage": "```js\n{\n\"_\": \"projectDataError:projectUnknownError\"\n}\n```",
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:3437]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 20]
```diff
@@ -3488,12 +3459,7 @@
         },
         "fields": {
             "description": "An array of fields defining annual crop management practices.",
-            "errorMessage": {
-                "_": "projectDataError:projectFieldsUnknownError",
-                "maxItems": "projectDataError:projectFieldsMaximumItemsError",
-                "minItems": "projectDataError:projectFieldsMinimumItemsError",
-                "type": "projectDataError:projectFieldsTypeError"
-            },
+            "errorMessage": "```js\n{\n\"minItems\": \"projectDataError:projectFieldsMinimumItemsError\",\n\"maxItems\": \"projectDataError:projectFieldsMaximumItemsError\",\n\"type\": \"projectDataError:projectFieldsTypeError\",\n\"_\": \"projectDataError:projectFieldsUnknownError\"\n}\n```",
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:3462]: this looks unexpected

File: packages/project/src/json/v4-specification.json [hunk 21]
```diff
@@ -3533,10 +3499,7 @@
         },
         "version": {
             "description": "The specification version. This information is used to determine the logic Nori uses to import a project",
-            "errorMessage": {
-                "_": "projectDataError:projectVersionUnknownError",
-                "type": "projectDataError:projectVersionTypeError"
-            },
+            "errorMessage": "```js\n{\n\"type\": \"projectDataError:projectVersionTypeError\",\n\"_\": \"projectDataError:projectVersionUnknownError\"\n}\n```",
```
Comment (jaycenhorton) [inline packages/project/src/json/v4-specification.json:3502]: this looks unexpected

File: packages/project/src/v3-specification.ts [hunk 22]
```diff
@@ -199,20 +199,24 @@ export const limingTypes = [
  * ```
  *
  * @errorMessage
+ * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v3-specification.ts:202]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v3-specification.ts [hunk 23]
```diff
@@ -199,20 +199,24 @@ export const limingTypes = [
...
  * }
+ * ```
  *
  */
 export interface Project {
   /**
    * The specification version. This information is used to determine the logic Nori uses to import a project
    *
    * @errorMessage
+   * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v3-specification.ts:214]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v3-specification.ts [hunk 24]
```diff
@@ -226,12 +230,14 @@ export interface Project {
    * An array of fields defining annual crop management practices.
    *
    * @errorMessage
+   * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v3-specification.ts:233]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v3-specification.ts [hunk 25]
```diff
@@ -550,11 +556,13 @@ export interface HistoricCRPLandManagement extends HistoricLandManagement {
  * ```
  *
  * @errorMessage
+ * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v3-specification.ts:559]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v3-specification.ts [hunk 26]
```diff
@@ -619,10 +627,12 @@ export interface Field {
    * ```
    *
    * @errorMessage
+   * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v3-specification.ts:630]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v3-specification.ts [hunk 27]
```diff
@@ -1271,10 +1281,12 @@ export interface CropEvent {
    * @validationRules ["cropEventDateIsOnOrAfterContainingCropYear"]
    *
    * @errorMessage
+   * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v3-specification.ts:1284]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v4-specification.ts [hunk 28]
```diff
@@ -312,20 +311,24 @@ export const limingTypes = [
  * ```
  *
  * @errorMessage
+ * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v4-specification.ts:314]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v4-specification.ts [hunk 29]
```diff
@@ -312,20 +311,24 @@ export const limingTypes = [
...
  * }
+ * ```
  *
  */
 export interface Project {
   /**
    * The specification version. This information is used to determine the logic Nori uses to import a project
    *
    * @errorMessage
+   * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v4-specification.ts:326]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v4-specification.ts [hunk 30]
```diff
@@ -405,12 +408,14 @@ export interface Project {
    * An array of fields defining annual crop management practices.
    *
    * @errorMessage
+   * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v4-specification.ts:411]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v4-specification.ts [hunk 31]
```diff
@@ -902,11 +907,13 @@ export interface PracticeChangesAdopted {
  * ```
  *
  * @errorMessage
+ * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v4-specification.ts:910]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v4-specification.ts [hunk 32]
```diff
@@ -1004,10 +1011,12 @@ export interface Field {
    * ```
    *
    * @errorMessage
+   * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v4-specification.ts:1014]: I think this is busting the schema.json that is generated (see comments)

File: packages/project/src/v4-specification.ts [hunk 33]
```diff
@@ -1147,6 +1164,7 @@ export interface Field {
  *  "crops": [
  *    // ... crops that were planted in year 2000
  *  ],
+ *  "dataTimeFrame": GROWER_REPORTED,
```
Comment (jaycenhorton) [inline packages/project/src/v4-specification.ts:1167]: should this be the following (I don't know the answer)
```suggestion
 *  "dataTimeFrame": "GROWER_REPORTED",
```

File: packages/project/src/v4-specification.ts [hunk 34]
```diff
@@ -1136,6 +1145,14 @@ export interface Field {
    */
   id?: string;
 }
+/**
+ * Enum indicating whether data is historical data reported by the grower,
+ * or a projection of future data.
+ */
+enum DataTimeFrame {
+  GrowerReported = 'GROWER_REPORTED',
+  Projected = 'PROJECTED',
```
Comment (jaycenhorton) [inline packages/project/src/v4-specification.ts:1154]: can we just keep the value and key the same?
```suggestion
  GROWER_REPORTED = 'GROWER_REPORTED',
  PROJECTED = 'PROJECTED',
```

File: packages/project/src/v4-specification.ts [hunk 35]
```diff
@@ -1775,10 +1807,12 @@ export interface CropEvent {
    * @format date
    *
    * @errorMessage
+   * ```js
```
Comment (jaycenhorton) [inline packages/project/src/v4-specification.ts:1810]: I think this is busting the schema.json that is generated (see comments)

**Review by jaycenhorton — CHANGES_REQUESTED** [review]

**Review by jaycenhorton — CHANGES_REQUESTED** [review]
Looked into the error message formatting issue pretty deep and my take-away is that we should:

1. revert the change that added enclosing code block tags (this breaks validation. Error messages are a custom tag that we use and the validation library requires that the contents of `@errorMessage` is an unescaped, non-code-blocked object (as it currently is on master.
2. stop emitting the error message bits into the docs as empty blocks, we should just exclude them when we run typedoc. Emitting the errorMessage tags as-is is not helpful in the docs anyways and is probably better documented in the errors package anyways since each error message in the spec should have a corresponding error definition in the error pacakge anyways. You can tell typedoc to exclude a tag [like this, using excludeTags](https://typedoc.org/guides/options/#excludetags)

File: package.json [hunk 36]
```diff
@@ -77,5 +77,10 @@
         "preset": "ts-jest"
       }
     ]
+  },
+  "dependencies": {
+    "project": "^0.1.6",
+    "typedoc": "0.23.17",
+    "typedoc-plugin-markdown": "3.13.6"
```
Comment (jaycenhorton) [inline package.json:84]:
```suggestion
```

Comment (jaycenhorton) [reply package.json:84]: the latter-two dependencies can be bumped in the project pacakge.json if needed

**Comment by ncanning** [comment]
I don't think typedoc likes `--excludeTags errorMessage`
```
$ /Users/nathan/code/nori/@nori-dot-com/nori-dot-com/node_modules/.bin/typedoc --readme none --excludeInternal --excludeExternals --excludeTags errorMessage --plugin typedoc-plugin-markdown src/v4-specification.ts src/v3-specification.ts src/validation.ts src/utils/convertFromV2ToV1.ts
Loaded plugin typedoc-plugin-markdown
error excludeTags must be an array of valid tag names.
error Command failed with exit code 1.
```
I'm guessing this is maybe because like you said `errorMessage` is a custom tag, so not something that's recognized by typedoc? Not really sure though

File: packages/project/src/v4-specification.ts [hunk 37]
```diff
@@ -1136,6 +1145,14 @@ export interface Field {
    */
   id?: string;
 }
+/**
+ * Enum indicating whether data is historical data reported by the grower,
+ * or a projection of future data.
+ */
+enum DataTimeFrame {
```
Comment (scottynomad) [inline packages/project/src/v4-specification.ts:1152]: `dataSourceType`

File: packages/project/src/json/v4-specification.json (hunk 8 above)
Comment (scottynomad) [reply packages/project/src/json/v4-specification.json]: removed the js comment wrapper around errorMessage annotations.

File: packages/project/src/validation.ts [hunk 38]
```diff
@@ -19,7 +18,7 @@ type ProjectDataValidationContext = Omit<DataValidationCxt, 'rootData'> & {
  *
  * @internal
  */
-export const formatInputData = (data: Project): Project => {
+export const formatInputData = (data: any) => {
```
Comment (mrdave-dev) [inline packages/project/src/validation.ts:21]: Intentional change?

**Review by mrdave-dev — APPROVED** [review]

File: package.json [hunk 39]
```diff
@@ -76,5 +77,10 @@
         "preset": "ts-jest"
       }
     ]
+  },
+  "dependencies": {
+    "project": "^0.1.6",
```
Comment (jaycenhorton) [inline package.json:82]: this isn't installing the dependency you think it is. [it's installing this 9 year old pacakge](https://www.npmjs.com/package/project)

**Review by jaycenhorton — CHANGES_REQUESTED** [review]

File: package.json [hunk 40]
```diff
@@ -11,6 +11,7 @@
     "jest": "^29.2.1",
     "lerna": "^5.1.6",
     "ts-jest": "^29.0.3",
+    "tsc-watch": "^4.5.0",
```
Comment (jaycenhorton) [inline package.json:14]: we don't need this anymore (it was removed [here](https://github.com/nori-dot-eco/nori-dot-com/pull/209))

Comment (jaycenhorton) [inline package.json:14]:
```suggestion
```

File: packages/project/src/__tests__/validation.test.ts [hunk 41]
```diff
@@ -44,36 +56,37 @@ const BASIC_UNFORMATTED_VALID_PROJECT: Project = {
...
-              limingEvents: null,
-              grazingEvents: null,
-              burningEvent: null,
-              soilOrCropDisturbanceEvents: [],
+              //   organicMatterEvents: [],
+              //   irrigationEvents: [],
+              //   limingEvents: null,
+              //   grazingEvents: null,
+              //   burningEvents: null,
+              //   soilOrCropDisturbanceEvents: [],
```
Comment (jaycenhorton) [inline packages/project/src/__tests__/validation.test.ts:85]:
```suggestion

```

File: packages/project/src/__tests__/validation.test.ts [hunk 42]
```diff
@@ -161,7 +180,7 @@ const BASIC_UNFORMATTED_INVALID_PROJECT: Project = {
     },
   ],
 };
-const clone = (object: Project): Project => JSON.parse(JSON.stringify(object));
+const clone = (object: Object) => JSON.parse(JSON.stringify(object));
```
Comment (jaycenhorton) [inline packages/project/src/__tests__/validation.test.ts:183]: Can you change this back to Project or use a better type?

Avoid the `Object` and `{}` types, as they mean "any non-nullish value". See [this comment for more information](https://github.com/typescript-eslint/typescript-eslint/issues/2063#issuecomment-675156492).

eslint should have caught this one. IF you intentionally meant for this to be changes, and meant any object type, use `Record<string, any>`

File: package.json (hunk 40 above)
Comment (scottynomad) [reply package.json:14]: Hmm, I must have messed up the merge, fixed.

**Review by jaycenhorton — APPROVED** [review]

File: packages/project/src/__tests__/validation.test.ts (hunk 42 above)
Comment (scottynomad) [reply packages/project/src/__tests__/validation.test.ts:183]: Yep, sorry, I undid that when I was debugging the tests, thanks for catching that.

**Pull request merged**